### ISSUE TITLE
Feature request: Ability to specify a secondary ship type

### ISSUE THREAD

**Issue by Tadrix** [endless-sky/endless-sky#2450]
Right now we have an ability to have ship variants by putting a secondary model name in the data file something like this:
```python
ship "Kestrel" "Kestrel (More Shields)"
```
What I ask for is an ability to specify a secondary ship type, like this:
```python
ship "Star Queen"
	sprite "ship/star queen"
	attributes
category "Transport" "Heavy Transport"
```

Benefits:
1. Future proofing. At one point there probably will be enough ships of the same type to sub-divide them.
I'm thinking about giving a "Heavy Transport" sub-type to Star Queen, a "Medium Transport" to Blackbird and a "Light Transport" sub-type to other transports, and a "Medium Freighter" sub-type to "Hauler II" and "Water Bug". Missions can be made taking these sub-types into account (checking, for example, if medium freighters + heavy freighters + heavy transports + heavy warships > 6 in conditions), so it won't be necessary to change them later if new ship types become visible in shipyards.
(By the way, I think Star Barge really is an ultralight freighter, Blackbird is a light transport and other transports also are ultralight, something similar to what an interceptor is to warships; I'd call these types "Launch", "Courier" or "Trader". I tried to leave things simple in my example.)
2. Ability for missions to better take into account what ships player's fleet utilizes. Because 3 Star Queens != 3 Shuttles. After all, we are not yet using those sub-types as types only because there are too ships in them and as a result shipyard's view would be needlessly cluttered. No reason to refrain from using them internally already.

**Comment by KiLEdEnNis** [comment]
I am for dividing stuff to Light/Medium/Heavy Transport/Freighter too, but this issue seems to be about being able to let missions require a certain ship (type)
why not just ask for being able to require a specific ship or ship type instead? since you kinda confused subdividing ship types into it
you could give a mission the requirement for at least 1 of several specific ships this way without needing sub-categories

**Comment by jafdy** [comment]
It would make more sense to allow the same ship to count twice. As well as being easier to code.

**Comment by KiLEdEnNis** [comment]
then maybe better coding for an "or" and "and" option; have 1 ship count only once or several times

**Comment by EndrosG** [comment]
The `or` and `and` operators already exist in conditions. So, since we talk about conditions (mainly for `to offer`, I guess) you can assume them to exist.

**Comment by Tadrix** [comment]
> to let missions require a certain ship (type)

If you have to specify ships exactly, you would need to revise the mission every time game gets new ships, and this would not work for plugins. My point is, if the mission (in vanilla) requires a "Heavy Transport" (and not a Star Queen specifically), and the plugin player uses features another "Heavy Transport", and the player happens to have such transport, then condition should be satisfied (in most cases it doesn't matter which high-capacity ship player uses) — making such missions **plugin-friendly**.

Furthermore, my opinion is that sub-types should internally be treated like separate types and just be shown to the player together in shipyard. _And_ it would be nice to be able to require ship type `Transport` (all sizes of transports), `Freighter` (all sizes of freighters), `Warship` (all sizes of warships, including Interceptors, excluding fighters and drones).

In other words, sub-types could also be used as semi-separate types which consist of too few ships to make a distinction yet (so they are shown to the player under encompassing over-type), but in fact are ready to be made separate at any time with as few necessary changes as possible (thus "**future-proofing**").

In other words, there would be more benefits and no drawbacks to use some kind of hierarchy among ship types, and be able to use these intermediary types in conditions and statistics.

**Comment by jafdy** [comment]
Other than code time.

**Comment by EndrosG** [comment]
@Tadrix : it looks like you have specified a 3-layer hierarchy in your last comment.

major category | visible category | sub category
-------------|-----------------|------------------
transport | transport | light transport
transport | transport | heavy transport
warship | interceptor | interceptor
warship | light warship | light warship

That seems more complex than in your first post.

+reference to pull #2200

**Comment by Tadrix** [comment]
I know, but it was just one way of doing what I proposed, which I thought would be beneficial to us. In other words, I recently got an idea that a lot of stuff (hulls, outfits, missions — to allow them to explicitly define as belonging to specific mission chain) could be put in a tree-like hierarchy. I see great potential to this. Unfortunately, I'm not nearly good enough at C++ to even attempt something like else on my own. I would be ready to rewrite (or help to rewrite) data files (*.txt) to use such a system, if someone took care of programming side of this thing. The best would be if these hierarchies could be defined in data too (no hard-coded ship types, for example, but specified in separate data file).

However, it isn't really necessary to my OP. OP is concerned mainly with ability to use what eventually would become a ship type.

For example, people generally agree that transports could be divided to heavy, medium and light transports, yet they also have opinion that it isn't necessary because there would be too few ships in each type and shipyard view would become too messy (which is true). However, one day we might need such distinction. What I proposed would allow to specify a secondary group for the ship, and then missions could refer to them by type, of by sub-type. When new type came into use, we wouldn't need to touch missions which dealt with the sub-type, we just would erase the former type.

Or, we would switch them around, so missions which refer to "transports" still work. This would be like a keyword system, where first value also determines how to show that ship in shipyard.

In other words, in keyword system, I would compare ship type to planet's attributes, the main one is first one (used by the shipyard), the additional ones are optional, yet ship could be referred to by them. For example, we could give Gunboat, Rainmaker, Frigate, Cruiser and Carrier an attribute ```Navy```. Not sure if this would be useful, though.

(Tree hierarchies' only difference would be that the game itself understands that "heavy transport" is also a "transport", you wouldn't need to mention it additionally in the data file, but, as I said, it's just one of ways to do things. For descriptive purposes we could also have fields "Manufacturer" or "Origin" if we needed them. Yeah, I probably was doing too much OOP-related stuff some time ago. xD)

**Comment by EndrosG** [comment]
Manufacturer / Origin is something I'd like to have.
Because it would allow for missions / spaceport evens reacting on alien ships.
E.g. with the new syMajor planet attribute introduced in #2449 and an manuf./origin attribute in the ship entity, we could create missions like
```
to offer
  has "some wanderer ship"
source
  government Republic "Free Worlds" Syndicate Neutral
  attributes syMajor
on offer
  conversation
    `Hey, I see you have some alien ship. Would you be willing to let our engineers onboard for a week or so? Of course it may happen that they use their screwdrivers and take it apart...`
```

**Comment by Elyssaen** [comment]
@EndrosG, I have an enhancement sat about gathering dust that allows this. Here's the commit description:

**[feature/TechConditions](https://github.com/Elyssaen/endless-sky/tree/feature/TechConditions)**
Content creators can give any outfit, weapon or ship "tech: " tags,
i.e., any attribute that begins with the string "tech: ". A number isn't
required, but can be optionally added. All tech tags you have on
outfits/ships in your active fleet are kept up-to-date in conditions, so
can be used for mission logic. e.g. Quarg outfits and ships could be
given "tech: quarg" and missions could trigger off having any of them.

____________________

I plan to go back to it at some point and remove the "tech: " aspect and replace it with a top level "tags" line and then just indented phrases for each tag.

**Comment by Tadrix** [comment]
And why not `has "hyperdrive" + "scram  drive" + "jump drive" > 1`?

**Comment by EndrosG** [comment]
@Elyssaen : The description you provided here reads good. And I agree with you, `tags` sounds more generally usable than `origin`. Having it in a separate node with indent sounds good to me.
Would you allow these tags to be ship specific? I.e. copy them to the single instance? Or just have them in the master definition? (the weapon used in the final explosion is such a definition only thing)... But if tags were per instance, once eventually could even tag captured ships this way (vs. bought ships).

Regarding the spare hyperdrive you'd still need to check that it's not just 2 hyperdrives because it's 2 ships in the fleet.

**Comment by Elyssaen** [comment]
@Tadrix: last I checked (January), you can't use `require outfit` that way. And I can't remember if you can add up conditions like this, though I should check as it's relevant to a project of mine that uses conditions and random factors to model a background simulation war.

But the deeper reason was, I thought it would be a good idea to create tags on Hyperdrives and Jump Drives reflecting their capabilities. A Jump Drive is capable of acting like a Hyperdrive, so should have that tag too, and if a new outfit was introduced by someone that could also act like a Hyperdrive, it should have it as well. Or if it was inherent to the ship, etc.

@EndrosG: was looking at the master definition only, but your idea is worth some thought! I've definitely wanted for a while some way of tracking captured ships better for mission purposes. All I've got at the moment is sticking a 'Black Box' outfit on it.

The two hyperdrives being split across a fleet wasn't too much of a problem for my mission idea... even in that case, the player won't become stranded if they give up the hyperdrive on their ship. (The mission idea is in a planet that doesn't sell them, bring a spare one for a significant profit. In your scenario, worst comes to worst, you can still switch your control ship to the one that keeps the hyperdrive, leave and come back with another for your other ship.)

By the way, I'm not sure I saw you around when I was active back in January and earlier? I took a hiatus, and looking about again you seem really active. I'd like to get active again, and if I do I'd like to be in touch with you about stuff. That okay?

**Comment by EndrosG** [comment]
@Elyssaen : You cant write such a sum statement in the game right now.
But you might want to checkout pull #2383 (merged), pull #2456 (not yet merged) and these three examples:
+ pull #2461 (needs also pull #2449, but that is just a content update, no code update)
+ the Eireen plugin [in my repo] (https://github.com/EndrosG/ES-Plugins/tree/master/Eireen/data) - the "First Contact" conversation is the main one, below follow several parallel R&D paths, reading one is sufficient
+ the R&D path I currently completely rewrite in my primary plugin `Marsian Ancestors` [again in my repo] (https://github.com/EndrosG/ES-Plugins/tree/master/Marsian Ancestors/data)

They all do some different sort of "simulation" using the stuff of the above mentioned pulls #2383 and #2456.

**Comment by EndrosG** [comment]
@Elyssaen (sorry for the 2nd reply) : I wasn't here in january. I play ES since over a year but registered to github just lately to make some contributions.
Parts of my Marsian Ancestors plugin date back to march 2016, but it is currently under construction again after I started using blender and after some of my coding allows for new simulation deepth.
I would be really happy if you contact me on any questions or so. You shouldn't have trouble finding me on facebook, *if you want*.

Regarding the hyperdrives: I am not sure if it is a good idea to use the tag's construct for that at all. But if you want to use it, just do as you said, and assign all kind of hyperdrives, including jump drives, a common tag, and assign them individual tags. Then, due to the common tag being summed up automatically, @Tadrix would not need his `has "hyperdrive" + "scram drive" + "jump drive" > 1` attempt.

Sadly, I am about to change the definitions of my Metropolis Segment ships in a way that they require (due to their size) several hyperdrives (or jump drives) installed in order to be able to perform a jump at all. Most likely I will give the ship base definition `attributes "jump speed" -7.0 and "scram drive" -7.0`, i.e. if you install 36 hyper drives (each having 0.2 jump speed = 7.2 in total) or 24 jump drives (each having 0.3 jump speed, again 7.2 in total), you reach a positive jump speed and only then the ship can actually perform a jump. Because - looking at the code - a ship can't enter hyperspace as long as it's speed is greater than the jump speed. And since the minimum speed you can reach is zero, this is always greater than any negative number.
Yeah, I know, I do (and intend to do) strange things :-D

**Comment by endless-sky** [comment]
There will probably never come a time when there are enough different ship types *in one shipyard* to justify splitting them up into more sub-categories. So this would only potentially be useful for the shipyards map. And there, having many different categories would make it hard to find a particular ship if you didn't remember what category it was in.
